core/sr-io/src/offchain/http.rs

```diff
@@ -0,0 +1,398 @@
+
+/// Request method (HTTP verb)
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Method {
+	/// GET request
+	Get,
+	/// POST request
+	Post,
+	/// PUT request
+	Put,
+	/// PATCH request
+	Patch,
+	/// DELETE request
+	Delete,
+	/// Custom verb
+	Other(&'static str),
+}
+
+impl AsRef<str> for Method {
+	fn as_ref(&self) -> &str {
+		match *self {
+			Method::Get => "GET",
+			Method::Post => "POST",
+			Method::Put => "PUT",
+			Method::Patch => "PATCH",
+			Method::Delete => "DELETE",
+			Method::Other(m) => m,
+		}
+	}
+}
+
+#[cfg(feature = "std")]
+fn from_utf8(chunk: &[u8]) -> Option<&str> {
+	std::str::from_utf8(chunk).ok()
+}
+#[cfg(not(feature = "std"))]
+fn from_utf8(chunk: &[u8]) -> Option<&str> {
+	core::str::from_utf8(chunk).ok()
+}
+
+/// An HTTP request builder.
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Request<'a, T = &'static [&'static [u8]]> {
+	/// Request method
+	pub method: Method,
+	/// Request URL
+	pub url: &'a str,
+	/// Body of the request
+	pub body: T,
+	/// Deadline to finish sending the request
+	pub deadline: Option<Timestamp>,
+	/// Request list of headers.
+	headers: Vec<(Vec<u8>, Vec<u8>)>,
+}
+
+impl<T: Default> Default for Request<'static, T> {
+	fn default() -> Self {
+		Request {
+			method: Method::Get,
+			url: "http://localhost",
+			headers: Vec::new(),
+			body: Default::default(),
+			deadline: None,
+		}
+	}
+}
+
+impl<'a, T: Default> Request<'a, T> {
+	/// Create new Request builder with given URL.
+	pub fn new(url: &'a str) -> Self {
+		let mut req = Request::default();
+		req.url(url);
+		req
+	}
+
+	/// Change the URL of the request.
+	pub fn url(&mut self, url: &'a str) -> &mut Self {
+		self.url = url;
+		self
+	}
+
+	/// Set the body of the request.
+	pub fn body(&mut self, body: T) -> &mut Self {
+		self.body = body;
+		self
+	}
+
+	/// Add a header.
+	pub fn add_header(&mut self, name: &str, value: &str) -> &mut Self {
+		self.headers.push((name.as_bytes().to_vec(), value.as_bytes().to_vec()));
+		self
+	}
+
+	/// Set the deadline of the request.
+	pub fn deadline(&mut self, deadline: Timestamp) -> &mut Self {
+		self.deadline = Some(deadline);
+		self
+	}
+}
+
+impl<'a, 'b, T: IntoIterator<Item=&'b [u8]>> Request<'a, T> {
+	/// Send the request and return a handle.
+	///
+	/// Err is returned in case the deadline is reached
+	/// or the request timeouts.
+	pub fn send(self) -> Result<PendingRequest, ()> {
+		let meta = &[];
+
+		// start an http request.
+		let id = crate::http_request_start(self.method.as_ref(), self.url, meta)?;
+
+		// add custom headers
+		for (header_name, header_value) in &self.headers {
+			crate::http_request_add_header(
+				id,
+				from_utf8(header_name).expect("Header names are always Vecs created from valid str; qed"),
+				from_utf8(header_value).expect("Header values are always Vecs created from valid str; qed"),
```

Comment:
Instead of converting strings to vectors and back to strings (which is by the way is not 0-cost, since it involves validation that takes linear time) can we store `String` in `self.headers`?

Reply:
Oh, that's bummer. This clearly demonstrates mismatch between priorities of on-chain and off-chain code. I believe sooner or later off-chain would want to use `String`s anyway. 

I saw your proposed solution. The best thing is to avoid unsafe if we can. If we can't then I think the following solution will be a bit better:

1. introduce a struct, let's call it `Header`.
2. In order to construct an instance of `Header` you have to provide `name: &str` and `value: &str`.
3. the body of the constructor would do `to_bytes().to_vec()`
4. then we add another function that adds this header to the request.

IMO this is better since the code that make guarantees and the code that relies on those guarantees is physically closer to each other. Furthermore, the whole invariant is the almost the single purpose of this struct. These facts should make the proof better, more maintanable and should decrease chances of violating the invariant accidentally with all of this being zero-cost.

wdyt?